File: Python/disjointSets.py
class Node():
    def __init__(self, x):
        self.val = x
        self.parent = None
        self.h = 1

class disjointSet():
    def __init__(self):
        self.dic = {}

    def makeSet(self, x):
        self.dic[x] = Node(x)

    def findSet(self, x):
        node = self.dic[x]
        while node.parent != None:
            node = self.dic[node.parent]
        return node.val

    def union(self,x,y):
        rx = self.dic[self.findSet(x)]
        ry = self.dic[self.findSet(y)]
        if rx.h < ry.h:
            rx.parent = ry.val
        elif rx.h > ry.h:
            ry.parent = rx.val
        else:
            ry.parent = rx.val
            rx.h +=1


def main():
    s = disjointSet()
    s.makeSet(3)
    s.makeSet(7)
    s.makeSet(4)
    s.makeSet(5)
    s.makeSet(6)
    s.makeSet(20)
    s.makeSet(2)
    s.makeSet(9)
    s.makeSet(16)
    s.union(4,9)
    s.union(3,7)
    s.union(2,16)
    s.union(6,5)
    s.union(5,20)
    s.union(3,5)
    for key in s.dic:
        print(key, s.dic[key].parent)


if __name__ == '__main__':
    main()

File: Cpp/disjointSets.cpp
#include <iostream>
#include <map>
#include <vector>

using namespace std;

template<typename T>
class disjoint_set{
public:
    T val;
    disjoint_set<T> *parent;
    int h;
    
    disjoint_set(T x){
        this->val = x;
        this->parent = NULL;
        this->h = 1;
    }
    
    
    disjoint_set<T>* findSet(){
        disjoint_set<T> *node = this;
        while(node->parent != NULL) node = node->parent;
        return node;
    }
    
    void unite(disjoint_set<T> *y){
        disjoint_set<T> *rx = this;
        disjoint_set<T> *ry = y;
        rx =  rx->findSet();
        ry = ry->findSet();
        if(rx->h < ry->h) rx->parent = ry;
        else if (rx->h > ry->h) ry->parent = rx;
        else{
            ry->parent = rx;
            rx->h++;
        }
    }
    
};

int main(){
    map<int, disjoint_set<int> * > m;
    m[3] = new disjoint_set<int>(3);
    m[7] = new disjoint_set<int>(7);
    m[4] = new disjoint_set<int>(4);
    m[5] = new disjoint_set<int>(5);
    m[6] = new disjoint_set<int>(6);
    m[20] = new disjoint_set<int>(20);
    m[2] = new disjoint_set<int>(2);
    m[9] = new disjoint_set<int>(9);
    m[16] = new disjoint_set<int>(16);
    m[4]->unite(m[9]);
    m[3]->unite(m[7]);
    m[2]->unite(m[16]);
    m[6]->unite(m[5]);
    m[5]->unite(m[20]);
    m[3]->unite(m[5]);
    for (auto x: m){
        cout << x.first <<  " " <<  x.second->findSet()->val << endl;
    }
    return 0;
}